### ISSUE TITLE
Bug in gradient accumulation in PPO trainer

### ISSUE THREAD

**Issue by Uncomfy** [lvwerra/trl#583]
Hello! Thank you for this library, it made working with PPO a lot easier and helped me to understand it better :)

I've been trying to run my training with gradient accumulation, however I noticed that reward/loss curves were looking way worse than without it, so I decided to investigate.
According to [this](https://huggingface.co/docs/accelerate/usage_guides/gradient_accumulation) tutorial, when using `accelerator.accumulate` you need to do `optimizer.zero_grad()` at the same time as `optimizer.step()`, however the code in the PPO trainer does `zero_grad()` every `gradient_accumulation_steps`: https://github.com/lvwerra/trl/blob/31658b42637a9bffb985f61a59cb0039ef2ca624/trl/trainer/ppo_trainer.py#L718-L719

You can verify that this is causing an issue by increasing the amount of samples in dummy dataset in `tests/test_ppo_trainer.py` to 64:
https://github.com/lvwerra/trl/blob/31658b42637a9bffb985f61a59cb0039ef2ca624/tests/test_ppo_trainer.py#L162-L163

And removing `break`s in the test loops:
https://github.com/lvwerra/trl/blob/31658b42637a9bffb985f61a59cb0039ef2ca624/tests/test_ppo_trainer.py#L1130
https://github.com/lvwerra/trl/blob/31658b42637a9bffb985f61a59cb0039ef2ca624/tests/test_ppo_trainer.py#L1154

You will see that this test will fail. However, if you modify the code of `ppo_trainer.py` like so:
```python
                    # if (i % self.config.gradient_accumulation_steps) == 0:
                    #     self.optimizer.zero_grad()

                    train_stats = self.train_minibatch(
                        batch["logprobs"],
                        batch["values"],
                        batch["rewards"],
                        logprobs,
                        logits,
                        vpreds,
                        batch["masks"],
                    )
                    self.optimizer.zero_grad()
```

then the test will be successful.

**Comment by younesbelkada** [comment]
Thank you very much @Uncomfy for your time and investigation
I have managed to reproduce the bug and can confirm your proposed fix seems to be the right fix. I added you as a co-author of the solution. 
Note also that there is an attempt to refactor the entire gradient accumulation protocol in https://github.com/lvwerra/trl/pull/546

**Comment by Uncomfy** [comment]
Thank you @younesbelkada for a quick fix!

**Comment by younesbelkada** [comment]
Closing as https://github.com/lvwerra/trl/pull/546 got merged
